### DISCUSSION TITLE
Proposal : auto conversion from anonymous object to tuples

### DISCUSSION THREAD

**Discussion by mariusmg** [dotnet/csharplang#1416] [Language Ideas]
It would be very nice if some autoconversion could be done from a anonymous object to tuples based on 
the fields/property names. So the following code would work :

public (string content, DateTime updateDate) Get(int id)
{
			return this.Find().Where(x => x.Id == id).Select(note => new {content = note.Content, updateDate = note.UpdateDate}).FirstOrDefault();
}

**Comment by CyrusNajmabadi** [comment]
Tagging @jcouv .  Definitely an interesting suggestion.  

Although, @mariusmg Can you explain your scenario more?  Your example is not compelling to me.  Why would you not just write:

```c#
public (string content, DateTime updateDate) Get(int id)
{
  return this.Find().Where(x => x.Id == id).Select(
    note => (content: note.Content, updateDate: note.UpdateDate)).FirstOrDefault();
}
```

// and, in practice, because of tuple conversions, writing the following would also be fine:


```c#
public (string content, DateTime updateDate) Get(int id)
{
  return this.Find().Where(x => x.Id == id).Select(
    note => (note.Content, note.UpdateDate)).FirstOrDefault();
}
```

Both of these seem better than the anonymous-type route.

**Comment by DavidArno** [comment]
I'm with @CyrusNajmabadi here. At first glance, this seems a nice idea, but upon reflection it's hard to think of a use-case that can't just be solved by using a tuple instead of an anonymous type, thus avoiding the need to convert.

**Comment by yaakov-h** [comment]
I would argue that you could have an `IQueryable` provider that does not understand `ValueTuple`, but you could always do `.Select(note => new { note.Content, note.UpdateDate }).ToArray().Select(n => (n.Content, n.UpdateDate))` as a workaround.

**Comment by mariusmg** [comment]
@CyrusNajmabadi 
Both of your samples fail for me with  CS8143   "An expression tree may not contain a tuple literal". My compiler version is 2.7.0.62715 (db02128e)   (VS 15.6.4)

With which version of csc does that work for you ?!


@yaakov-h 
I get a predictable type conversion failure 
CS0029	Cannot implicitly convert type 'System.Collections.Generic.IEnumerable<(string Content, System.DateTime UpdateDate)>' to '(string content, System.DateTime udapteDate)'	

Regarding my scenario,  i have a bunch of small methods which do EF queries and return anonymous objects now (becase they only select a few columns from the table). I want to replace the return types of those methods with tuples (personally i think the resulting code would be simpler and better ). It would be nice if the resulting tuple could be created directly from the query, instead of the workaround with a intermediary anonymous object which has the query result and creating the resulting tuple from that.

**Comment by yaakov-h** [comment]
How do you have a method that return an anonymous object? Dynamic?

**Comment by mariusmg** [comment]
The method returns the tuple itself (first post)

public (string content, DateTime updateDate) Get(int id)

**Comment by jnm2** [comment]
> I would argue that you could have an `IQueryable` provider that does not understand `ValueTuple`, but you could always do `.Select(note => new { note.Content, note.UpdateDate }).ToArray().Select(n => (n.Content, n.UpdateDate))` as a workaround.

You're right, but I am \*so tired\* of doing this.

**Reply by Trolldemorted** [reply]
is there a remedy for this by now?

**Reply by CyrusNajmabadi** [reply]
@Trolldemorted there is not.

**Comment by CyrusNajmabadi** [comment]
I'd rather just have expression trees support tuples versus adding an anonymous-type-to-tuple conversion.

**Comment by alrz** [comment]
>I'd rather just have expression trees support tuple

At this point, it seems to me that we're waiting for the syntax to somehow "stabilizes" se we can go ahead and add expression tree support for all that in one go. I hope it becomes something like IOperation so every language feature comes with its own expression tree support. However, since those types are considered as an external dependency, it may be a burden to gradually add expression tree support for each feature.

**Comment by mariusmg** [comment]
> I'd rather just have expression trees support tuples

Sounds great. Anything that helps tuples feel more like first class citizens :)

**Comment by CyrusNajmabadi** [comment]
@jcouv Would it be on you to have expression trees support tuples?

**Comment by jcouv** [comment]
I don't know. I think we will have this discussion in C# 8.0 context, but nothing concrete yet.

**Comment by sab39** [comment]
I had a scenario where this bit me recentlly: using EF, I had various queries that all started with
```cs
  from x in db.Table1s
  where [complicated condition 1]
  from y in db.Table2s
  where [complicated condition 2]
```
and then continued differently from there. I wanted to refactor out the common part into a shared helper method, but there was no way to do it: Linq translates the above into an implicit select of anonymous type `{ Table1 x, Table2 y }`, so ideally the helper would return an `IQueryable` of that type, but you can't write a method with an unspeakable return type. The natural solution would be to return a tuple `(Table1 x, Table2 y)`, but that fails for lack of tuple support in expression trees.

I don't think it's particularly controversial to say that in an ideal world the solution would be to enhance expression trees to support tuples and other new language features, but per @MadsTorgersen [here](https://github.com/dotnet/csharplang/discussions/6526#discussioncomment-3851250), the amount of work involved is disheartening, which suggests the odds of getting it in C#12 are low. It'd certainly be nice if there were _some_ way to address the pain point in the short term.

One possibility would be to make the names of anonymous types speakable. I'm not sure what the original reason was for not allowing this, but if the concern was verbosity, the speakability of tuple types suggests it would be less of an obstacle today. A possible syntax might be something like `@{ Table1 x, Table2 y }` or even `(class { Table1 x, Table2 y })`. The latter has the downside that extra parentheses are (I think) required to avoid ambiguity when parsing, but on the other hand, would open up the option for future features like `(enum { X, Y })` for anonymous enums.

Making anonymous types speakable would be a separate discussion from this one though. Does such an issue already exist?

**Reply by HaloFour** [reply]
> I'm not sure what the original reason was for not allowing this, but if the concern was verbosity, the speakability of tuple types suggests it would be less of an obstacle today.

The problem is that speakability now requires that the C# specification now has to have very specific and consistent rules that describe how to generate the type name as if it's referenced from anywhere else that names has to be predictably recreated by any and all compiler implementations.  It's a massive amount of work, and if you need a contract of that nature then you can always declare your own type rather than using anonymous projections.

**Reply by sab39** [reply]
That makes sense. There might be room for a middle ground if the generated type was treated as `internal` so that it couldn't be exposed to external code anyway (although `[InternalsVisibleTo]` might present some gotchas). That might be a "worst of both worlds" thing though as I imagine a feature that has (seemingly) arbitrary limitations might lead to more frustration than a feature that doesn't exist at all.

My reason for not creating a class of my own was that I assumed (maybe wrongly?) that EF's expression translation to SQL is aware of anonymous types and gives them special behavior that a hand-built class would not get, because it would have to allow for property setters and getters that have side effects. If that's not the case then my pain point goes away!

**Comment by Rekkonnect** [comment]
I cannot imagine a reason to support this idea. I'm more of a fan of strongly-typed strucutres that do not include magic values specifically used in oneshots. I'd rather use the newly introduced `file` modifier on that oneshot type and explicitly declare it with whatever I want to operate on. Anonymous types are for very short-lived operations like delcaring what to select from a LINQ query (the main reason they got introduced) where you would really hate having to explicitly declaring it. Anything outside that scope requires that you have strongly-typed information about the object you receive and use, which delegates that responsibility back to having a fully-structured and declared type that you can use.
